### PR TITLE
smart loading for the routine library [POFSP-844]

### PR DESCRIPTION
**Pull request by polomani**
Only load items outside of already fetched time range (same as done for model library). This becomes more crucial now, since we only support 20 items per page for the routine revision list endpoint.

We sort of had this behavior back in the day, but then there were a lot of changes related to routines, so it was hard to keep it around.
Now it's time to use it again.

### CODE REVIEW

File: Cognite.Simulator.Tests/UtilsTests/TestUtilities.cs [hunk 1]
```diff
@@ -83,13 +83,13 @@ public static void WriteConfig()
...
         private static HttpResponseMessage ForbiddenResponse =
-            CreateResponse(HttpStatusCode.Forbidden, "{\"error\": {\"code\": 403,\"message\": \"Forbidden\"}}");
+            CreateResponse(HttpStatusCode.Gone, "{\"error\": {\"code\": 410,\"message\": \"Gone\"}}");
 
         /// <summary>
         /// Mocks the HttpClientFactory to return the mocked responses
         /// Example format for endpointMappings:
         ///     { uri => uri.Contains("/extpipes"), (MockExtPipesEndpoint, 0, 2) },
-        ///     2 is the number of times the response function will be called before returning a 403 Forbidden
+        ///     2 is the number of times the response function will be called before returning a 410 Gone
```
Comment (polomani) [inline Cognite.Simulator.Tests/UtilsTests/TestUtilities.cs:92]: this is just to differentiate from an actual 403. looks less serious in the logs

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs [hunk 2]
```diff
@@ -191,21 +200,38 @@ protected virtual V LocalConfigurationFromRoutine(SimulatorRoutineRevision routi
...
+            });
 
-            return localConfiguration;
+            return result;
         }
 
-        private async Task ReadRoutineRevisions(CancellationToken token)
+        private async Task ReadRoutineRevisions(bool init, CancellationToken token)
         {
+            string timeRange = LibState.DestinationExtractedRange.IsEmpty ? "Empty" : LibState.DestinationExtractedRange.ToString();
```
Comment (abdullah-cognite) [inline Cognite.Simulator.Utils/RoutineLibraryBase.cs:217]: Could this (`LibState.DestinationExtractedRange`) ever be `null`?

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs [hunk 3]
```diff
@@ -191,21 +200,38 @@ protected virtual V LocalConfigurationFromRoutine(SimulatorRoutineRevision routi
...
+            _logger.LogDebug("Updating routine library. There are currently {Num} routine revisions. Extracted range: {TimeRange}",
```
Comment (abdullah-cognite) [inline Cognite.Simulator.Utils/RoutineLibraryBase.cs:218]: Should we also add the `Init` flag to the debug message?

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs [hunk 4]
```diff
@@ -215,12 +241,21 @@ private async Task ReadRoutineRevisions(CancellationToken token)
...
-            foreach (var routineRev in routineRevisions)
+            if (routineRevisions.Any())
             {
-                if (!RoutineRevisions.ContainsKey(routineRev.Id.ToString()))
+                foreach (var routineRev in routineRevisions)
                 {
                     ReadAndSaveRoutineRevision(routineRev);
                 }
+
+                var maxCreatedMs = RoutineRevisions
```
Comment (abdullah-cognite) [inline Cognite.Simulator.Utils/RoutineLibraryBase.cs:251]: nitpick : can we rename the variables to something like ? 

`maxCreatedMs` -> `maxCreatedTimestamp` 

`maxCreatedDt ` -> `maxCreatedDateTime`

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs [hunk 5]
```diff
@@ -215,12 +241,21 @@ private async Task ReadRoutineRevisions(CancellationToken token)
...
+                    .Select(s => s.Value.CreatedTime)
+                    .Max();
+
+                var maxCreatedDt = CogniteTime.FromUnixTimeMilliseconds(maxCreatedMs);
+                LibState.UpdateDestinationRange(
```
Comment (abdullah-cognite) [inline Cognite.Simulator.Utils/RoutineLibraryBase.cs:256]: Should we log this as well?

`_logger.LogDebug("Destination range updated to start: {Start}, end: {End}",
    maxCreatedDt, maxCreatedDt);`

**Review by abdullah-cognite — APPROVED** [review]
overall looks good

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs (hunk 2 above)
Comment (polomani) [reply Cognite.Simulator.Utils/RoutineLibraryBase.cs:217]: no,
it's in the constructor

```
    public BaseExtractionState(string id)
    {
        Id = id;
        DestinationExtractedRange = TimeRange.Empty;
    }
```

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs (hunk 3 above)
Comment (polomani) [reply Cognite.Simulator.Utils/RoutineLibraryBase.cs:218]: happens only once. will clutter the logs

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs (hunk 5 above)
Comment (polomani) [reply Cognite.Simulator.Utils/RoutineLibraryBase.cs:264]: OK

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs (hunk 4 above)
Comment (polomani) [reply Cognite.Simulator.Utils/RoutineLibraryBase.cs]: OKI

File: Cognite.Simulator.Utils/RoutineLibraryBase.cs (hunk 3 above)
Comment (polomani) [reply Cognite.Simulator.Utils/RoutineLibraryBase.cs]: sure

**Pull request merged**